### PR TITLE
feat(plugin-cc): added-call-control

### PR DESCRIPTION
**Pull request by adhmenon**
# COMPLETES #< [SPARK-562170 (Ticket with Logs Attached](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-562170)>
# COMPLETES #< [SPARK-562180](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-562180)>
# COMPLETES #< [SPARK-569541](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-569541)>
# COMPLETES #< [SPARK-562179](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-562179)>

## This pull request addresses

* Introduction of call control actions such as hold, resume, end & wrapup call.
* Addition of buttons for the above in the sample app.
* It is currently based against - https://github.com/webex/webex-js-sdk/pull/4016
* Will migrate it once the above PR is merged.

## by making the following changes

* In the `Task` class, we call the respective APIs for hold, resume, end and wrapup.
* Added the required UI in sample app and made changes to incorporate them.
* Unit tested thoroughly

Vidcast - https://app.vidcast.io/share/c2b25b2c-d316-4e41-bbcd-26cced6bc863

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios were tested

* Tested all flows including registration, login and logout to ensure we are getting the call and call controls are [working](https://app.vidcast.io/share/59a82ceb-89ed-44da-a8ac-5b245839d533)
* Tested with reloading the webpage and checking if [relogin with call control was working](https://app.vidcast.io/share/0a79e410-21c3-4e83-91ed-c4e9280fff03)
* Tested by switching of network, waiting for [web socket reconnect and then checking if call control functionality was working](https://app.vidcast.io/share/6e7d45f7-5a9b-4617-b9d3-9aebc875acf3)
* Tested by ending call on external user end and [checking for wrapup enable](https://app.vidcast.io/share/7249f4a8-b5ad-45e7-baae-f3b480660cb3)
* Tested with [AGENT_EXTENSION](https://app.vidcast.io/share/6380f43e-6ad9-41e5-bf65-f53598c37c33)
* Agent_DN [testing call control](https://app.vidcast.io/share/a90a4728-96ad-47f1-8a12-d2e8b200e9d3)

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: docs/samples/contact-center/app.js [hunk 1]
```diff
@@ -21,6 +21,7 @@ let agentId;
 let taskControl;
 let task;
 let taskId;
+let agentAuxCodeId;
```
Comment (adhmenon) [inline docs/samples/contact-center/app.js:24]: We need to store this so that we can inject it into the `wrapup` method.

File: docs/samples/contact-center/app.js [hunk 2]
```diff
@@ -293,6 +299,10 @@ incomingCallListener.addEventListener('task:incoming', (event) => {
 function answer() {
   answerElm.disabled = true;
   declineElm.disabled = true;
+  holdElm.disabled = false;
```
Comment (adhmenon) [inline docs/samples/contact-center/app.js:302]: Had to make this change here as I was not getting the other event even after accepting the call.
If this is wrong, will change it.

File: docs/samples/contact-center/app.js [hunk 3]
```diff
@@ -361,3 +371,51 @@ function expandAll() {
   });
 }
 
+function holdCall() {
```
Comment (adhmenon) [inline docs/samples/contact-center/app.js:374]: Methods for all call control along with logs and error catches.

File: packages/@webex/plugin-cc/src/cc.ts [hunk 4]
```diff
@@ -23,7 +23,7 @@ import {StateChange, Logout} from './services/agent/types';
 import {getErrorDetails} from './services/core/Utils';
 import {Profile, WelcomeEvent} from './services/config/types';
 import {AGENT_STATE_AVAILABLE} from './services/config/constants';
-import {ConnectionLostDetails} from './services/core/websocket/types';
+import {ConnectionLostDetails} from './services/core/WebSocket/types';
```
Comment (adhmenon) [inline packages/@webex/plugin-cc/src/cc.ts:26]: For some reason, this was not letting me build and coming as an error, so I added this change. Will update once Priya's PR goes in.

File: packages/@webex/plugin-cc/src/services/core/WebSocket/connection-service.ts [hunk 5]
```diff
@@ -53,7 +53,7 @@ export class ConnectionService extends EventEmitter {
       isKeepAlive: this.isKeepAlive,
     };
     this.webSocketManager.handleConnectionLost(event);
-    LoggerProxy.log(`Dispatching connection lost event`, {
+    LoggerProxy.log(`Dispatching connection event`, {
```
Comment (adhmenon) [inline packages/@webex/plugin-cc/src/services/core/WebSocket/connection-service.ts]: This log was misleading, hence I modified it.

File: packages/@webex/plugin-cc/src/services/core/aqm-reqs.ts [hunk 6]
```diff
@@ -7,7 +7,7 @@ import {CbRes, Conf, ConfEmpty, Pending, Req, Res, ResEmpty} from './types';
 import {TIMEOUT_REQ} from './constants';
 import {AQM_REQS_FILE} from '../../constants';
 import HttpRequest from './HttpRequest';
-import {WebSocketManager} from './websocket/WebSocketManager';
+import {WebSocketManager} from './WebSocket/WebSocketManager';
```
Comment (adhmenon) [inline packages/@webex/plugin-cc/src/services/core/aqm-reqs.ts:10]: Nit: Same reason as mentioned above.

File: packages/@webex/plugin-cc/src/services/task/index.ts [hunk 7]
```diff
@@ -86,5 +86,92 @@ export default class Task implements ITask {
     }
   }
 
-  // TODO: Hold/resume, recording pause/resume, consult and transfer public methods to be implemented here
+  /**
+   * This is used to hold the task.
```
Comment (adhmenon) [inline packages/@webex/plugin-cc/src/services/task/index.ts:113]: Have added comments, documentation and methods for each call control. Note that the promise accept/reject is handled in aqm-reqs and we supply the events in the contacts.ts of the task folder (as done in Priya's PR). Everything works smoothly.

File: packages/@webex/plugin-cc/src/services/task/index.ts [hunk 8]
```diff
@@ -86,5 +86,92 @@ export default class Task implements ITask {
...
+   * @param taskId - Unique Id to identify each task
+   * @param data - WrapupPayLoad
+   * @returns Promise<TaskResponse>
+   * @throws Error
+   * @example
+   * ```typescript
+   * task.wrapup(taskId, data).then(()=>{}).catch(()=>{})
+   * ```
+   */
+  public async wrapup(taskId: string, wrapupPayload: WrapupPayLoad): Promise<TaskResponse> {
```
Comment (adhmenon) [inline packages/@webex/plugin-cc/src/services/task/index.ts:163]: I decided that we should accept the wrapup payload here - reason is it's very hard to get the agent config into this Task file, it is better if it is supplied by the external user. Further, the reason also needs to be supplied so I thought we can leave it open.
However, open to any other suggestions or changes.

File: packages/@webex/plugin-cc/test/unit/spec/services/Task/index.ts [hunk 9]
```diff
@@ -106,4 +110,98 @@ describe('Task', () => {
     await expect(task.decline(taskId)).rejects.toThrow(new Error(error.details.data.reason));
     expect(getErrorDetailsSpy).toHaveBeenCalledWith(error, 'decline', CC_FILE);
   }); 
+
+  it('should hold the task', async () => {
```
Comment (adhmenon) [inline packages/@webex/plugin-cc/test/unit/spec/services/Task/index.ts:114]: Just tested the functionality and the error handling.

**Comment by adhmenon** [comment]
Note - Currently I have set the destination as Priya's PR branch, will change once her PR goes in. Need to wait for one unit test fix on her PR also.

File: docs/samples/contact-center/app.js [hunk 10]
```diff
@@ -143,6 +148,7 @@ function register() {
         teamsDropdown.innerHTML = ''; // Clear previously selected option on teamsDropdown
         const listTeams = agentProfile.teams;
         agentId = agentProfile.agentId;
+        agentAuxCodeId = agentProfile.defaultWrapupCode;
```
Comment (rsarika) [inline docs/samples/contact-center/app.js]: This should be done using dropdown.

**Review by rsarika — APPROVED** [review]

File: docs/samples/contact-center/app.js (hunk 10 above)
Comment (adhmenon) [reply docs/samples/contact-center/app.js]: Noted. As discussed offline, I'll be adding a dropdown here which is populated with the `profile.wrapupcodes`. Thanks for the catch!

**Comment by adhmenon** [comment]
Added new dropdown for wrapup in call control as per Ravi's advice. Made a new vidcast and attached. Tested all other flows as well and everything is working fine.

File: docs/samples/contact-center/app.js [hunk 11]
```diff
@@ -42,6 +31,11 @@ const incomingDetailsElm = document.querySelector('#incoming-call');
 const answerElm = document.querySelector('#answer');
 const declineElm = document.querySelector('#decline');
 const callControlListener = document.querySelector('#callcontrolsection');
+const holdElm = document.querySelector('#hold');
+const resumeElm = document.querySelector('#resume');
+const endElm = document.querySelector('#end');
+const wrapupElm = document.querySelector('#wrapup');
+const wrapupCodesDropdownElm = document.querySelector('#wrapupCodesDropdown');
```
Comment (adhmenon) [inline docs/samples/contact-center/app.js:38]: We will now store the entire list of codes using this dropdown and populate the request accordingly.

File: docs/samples/contact-center/app.js [hunk 12]
```diff
@@ -189,6 +186,16 @@ function register() {
     }) 
 }
 
+function populateWrapupCodesDropdown() {
```
Comment (adhmenon) [inline docs/samples/contact-center/app.js:211]: We need name and id for the payload for wrapupm hence we store only that as of now.

File: packages/@webex/plugin-cc/src/services/core/aqm-reqs.ts [hunk 13]
```diff
@@ -98,12 +98,12 @@ export default class AqmReqs {
             if ('errId' in notifFail) {
               LoggerProxy.log(`Routing request failed: ${msg}`, {
                 module: AQM_REQS_FILE,
-                method: this.createPromise.name,
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/core/aqm-reqs.ts:101]: Why was this change required?

File: docs/samples/contact-center/index.html [hunk 14]
```diff
@@ -140,13 +140,26 @@ <h2 class="collapsible">
...
+            <audio id="remote-audio" autoplay></audio>
+          </fieldset>
         </div>
           <!-- Agent Call Controls -->
           <div id="callcontrolsection">
             <fieldset>
               <legend>Call Controls</legend>
               <div class="u-mv">
-                <!-- Add the call control buttons here -->
+                <button onclick="holdCall()" id="hold" class="btn--yellow" disabled>Hold</button>
```
Comment (Kesari3008) [inline docs/samples/contact-center/index.html]: We can use the same button also to do hold/resume

File: packages/@webex/plugin-cc/src/services/core/websocket/connection-service.ts [hunk 15]
```diff
@@ -53,9 +53,9 @@ export class ConnectionService extends EventEmitter {
       isKeepAlive: this.isKeepAlive,
     };
     this.webSocketManager.handleConnectionLost(event);
-    LoggerProxy.log(`Dispatching connection lost event`, {
+    LoggerProxy.log(`Dispatching connection event`, {
       module: CONNECTION_SERVICE_FILE,
-      method: this.dispatchConnectionEvent.name,
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/core/websocket/connection-service.ts:58]: same question here

File: packages/@webex/plugin-cc/src/services/task/index.ts [hunk 16]
```diff
@@ -86,5 +101,92 @@ export default class Task implements ITask {
...
+   * This is used to hold the task.
+   * @param taskId
+   * @returns Promise<TaskResponse>
+   * @throws Error
+   * @example
+   * ```typescript
+   * task.hold(taskId).then(()=>{}).catch(()=>{})
+   * ```
+   * */
+  public async hold(taskId: string): Promise<TaskResponse> {
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/task/index.ts]: We won't need the taskId as param to be passed here. Pull this information from this.data.interactionId to pas it to the contact layer

File: packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+      },
+    };
+
+    jest.spyOn(webCallingService, 'declineCall').mockImplementation(() => { throw error });
+    await expect(task.decline()).rejects.toThrow(new Error(error.details.data.reason));
+    expect(getErrorDetailsSpy).toHaveBeenCalledWith(error, 'decline', CC_FILE);
+  }); 
+
+  it('should hold the task', async () => {
+    await task.hold(taskId);
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts]: We should also add a check for a successful response from this api invocation, like what happens in then block. same comment applicable for other testcases

File: packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+      },
+    };
+    contactMock.end.mockImplementation(() => { throw error; });
+
+    await expect(task.end(taskId)).rejects.toThrow(error.details.data.reason);
+    expect(getErrorDetailsSpy).toHaveBeenCalledWith(error, 'end', CC_FILE);
+  });
+
+  it('should wrap up the task', async () => {
+    const wrapupPayload = {
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts:269]: Few test scenarios that we should cover manually as well as in UT:

1. What happens if we try to wrapup without ending the call?
2. Will wrapup succeed if wrapuppayload received from the user is empty strings? How is our public method handling it?
3. Initiate a call, end it from the customer side and do not wrapup. Initiate another call and see if we are able to handle it

File: packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+        },
+      },
+    };
+
+    jest.spyOn(webCallingService, 'declineCall').mockImplementation(() => { throw error });
+    await expect(task.decline()).rejects.toThrow(new Error(error.details.data.reason));
+    expect(getErrorDetailsSpy).toHaveBeenCalledWith(error, 'decline', CC_FILE);
+  }); 
+
+  it('should hold the task', async () => {
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts]: A test scenario to cover manually as well as in UT:

What happens in our CC-SDK if we do not pass the media resource ID in both cases hold and resume?

File: docs/samples/contact-center/app.js [hunk 20]
```diff
@@ -361,3 +375,54 @@ function expandAll() {
...
+    resumeElm.disabled = true;
+    wrapupElm.disabled = false;
+    wrapupCodesDropdownElm.disabled = false;
+  }).catch((error) => {
+    console.error('Failed to end the call', error);
+    endElm.disabled = false;
+  });
+}
+
+function wrapupCall() {
```
Comment (Kesari3008) [inline docs/samples/contact-center/app.js:457]: If the call is not ended from our kitchen sink app, then does the wrapup button become active and are we able to wrapup?

**Review by Kesari3008 — CHANGES_REQUESTED** [review]
Minor comments, more focus is on covering the test scnearios mentioned

File: packages/@webex/plugin-cc/src/services/core/aqm-reqs.ts (hunk 13 above)
Comment (adhmenon) [reply packages/@webex/plugin-cc/src/services/core/aqm-reqs.ts:101]: It was actually logging the wrong value - "value" instead of the "createPromise". I noticed it across a few other places. Fixed them in my current PR.

File: packages/@webex/plugin-cc/src/services/core/websocket/connection-service.ts (hunk 15 above)
Comment (adhmenon) [reply packages/@webex/plugin-cc/src/services/core/websocket/connection-service.ts:58]: Same reason as above.

File: packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts (hunk 19 above)
Comment (adhmenon) [reply packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts]: It fails to hold/resume... the call is continued as usual

File: packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts (hunk 18 above)
Comment (adhmenon) [reply packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts:254]: I can answer for a few - 
2. It won't work... the string has to match the wrapup aux code name. It'll throw an error in that case.
3. I am not sure about this.... will check.

File: docs/samples/contact-center/app.js (hunk 20 above)
Comment (adhmenon) [reply docs/samples/contact-center/app.js:457]: Resolved in latest commit.

File: docs/samples/contact-center/app.js [hunk 21]
```diff
@@ -293,6 +309,11 @@ incomingCallListener.addEventListener('task:incoming', (event) => {
 function answer() {
   answerElm.disabled = true;
   declineElm.disabled = true;
+  holdResumeElm.disabled = false;
+  holdResumeElm.innerText = 'Hold';
+  pauseResumeRecordingElm.disabled = false;
```
Comment (adhmenon) [inline docs/samples/contact-center/app.js:314]: Added pause and resume as well, along with a checkbox for `autoResume`.

File: packages/@webex/plugin-cc/src/services/task/index.ts [hunk 22]
```diff
@@ -86,5 +109,130 @@ export default class Task implements ITask {
...
+  /**
+   * This is used to pause the call recording
+   * @returns Promise<TaskResponse>
+   * @throws Error
+   * @example
+   * ```typescript
+   * task.pauseRecording().then(()=>{}).catch(()=>{});
+   * ```
+   */
+  public async pauseRecording(): Promise<TaskResponse> {
```
Comment (adhmenon) [inline packages/@webex/plugin-cc/src/services/task/index.ts:210]: Added pause and resume, tested it out on the updated vidcast.

File: packages/@webex/plugin-cc/src/services/task/TaskManager.ts [hunk 23]
```diff
@@ -51,18 +58,30 @@ export default class TaskManager extends EventEmitter {
...
+            this.emit(TASK_EVENTS.TASK_ASSIGNED, this.currentTask);
+            break;
+          case CC_EVENTS.CONTACT_ENDED:
+            this.emit(TASK_EVENTS.TASK_UNASSIGNED, this.currentTask);
+            if (this.webCallingService.loginOption === LoginOption.BROWSER) {
+              this.currentTask.unregisterWebCallListeners();
+              this.webCallingService.unregisterCallListeners();
+            }
+            break;
+          case CC_EVENTS.AGENT_WRAPUP:
```
Comment (adhmenon) [inline packages/@webex/plugin-cc/src/services/task/TaskManager.ts:82]: Need this in case user on other end drops off... there is an issue with the reading of events. 
Will sync and look into this issue.

File: packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts (hunk 18 above)
Comment (adhmenon) [reply packages/@webex/plugin-cc/test/unit/spec/services/task/index.ts:269]: Manually tested all flows, working as expected - 
1. Throws an error
2. Added a condition inside our public method to test this, it will now throw an error.
3. Handled. Please see https://app.vidcast.io/share/7249f4a8-b5ad-45e7-baae-f3b480660cb3

File: packages/@webex/plugin-cc/src/services/task/index.ts [hunk 24]
```diff
@@ -86,5 +109,144 @@ export default class Task implements ITask {
...
   }
 
-  // TODO: Hold/resume, recording pause/resume, consult and transfer public methods to be implemented here
+  /**
+   * This is used to hold the task.
+   * @returns Promise<TaskResponse>
+   * @throws Error
+   * @example
+   * ```typescript
+   * task.hold(taskId).then(()=>{}).catch(()=>{})
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/task/index.ts]: Please update tsdoc descriptions since taskId as param not required anymore

File: packages/@webex/plugin-cc/src/services/task/index.ts [hunk 25]
```diff
@@ -86,5 +109,144 @@ export default class Task implements ITask {
...
+   * @returns Promise<TaskResponse>
+   * @throws Error
+   * @example
+   * ```typescript
+   * task.resume(taskId).then(()=>{}).catch(()=>{})
+   * ```
+   */
+  public async resume(): Promise<TaskResponse> {
+    try {
+      if (!this.data.mediaResourceId || this.data.mediaResourceId.length === 0) {
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/task/index.ts]: If we know that empty mediaresourceId fails the hold/resume, would we still need this condition. Isn't contact layer handling this?

File: packages/@webex/plugin-cc/src/services/task/TaskManager.ts [hunk 26]
```diff
@@ -51,18 +60,34 @@ export default class TaskManager extends EventEmitter {
...
+            this.currentTask = this.currentTask.updateTaskData(payload.data);
+            this.currentTask.emit(TASK_EVENTS.TASK_END, this.currentTask);
+            break;
+          case CC_EVENTS.AGENT_WRAPPEDUP:
+            this.removeCurrentTaskFromCollection();
+            this.currentTask.emit(TASK_EVENTS.TASK_WRAPUP, this.currentTask);
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/src/services/task/TaskManager.ts]: Since this event indicates complete task wrapup, I believe we don't need to emit task object

File: packages/@webex/plugin-cc/test/unit/spec/services/task/TaskManager.ts [hunk 27]
```diff
@@ -0,0 +1,296 @@
...
+jest.mock('./../../../../../src/services/task', () => {
+  return jest.fn().mockImplementation(() => {
+    return {
+        updateTaskData: jest.fn().mockReturnThis(),
+        emit: jest.fn()
+    };
+  });
+});
+
+describe('TaskManager', () => {
```
Comment (Kesari3008) [inline packages/@webex/plugin-cc/test/unit/spec/services/task/TaskManager.ts]: Can we add UT for Agent wrapup event added where task is removed from the collection and also where end has happened from the customer end and taks: end is emitted

**Review by Kesari3008 — COMMENTED** [review]
Minor comments

File: packages/@webex/plugin-cc/src/services/task/index.ts (hunk 25 above)
Comment (adhmenon) [reply packages/@webex/plugin-cc/src/services/task/index.ts]: This is a valid point... the more I see it, I think we can forego this entire condition... it anyways will fail for an empty mediaResourceId.
Is that alright @Kesari3008?

File: packages/@webex/plugin-cc/test/unit/spec/services/task/TaskManager.ts (hunk 27 above)
Comment (adhmenon) [reply packages/@webex/plugin-cc/test/unit/spec/services/task/TaskManager.ts]: Added them.

File: packages/@webex/plugin-cc/src/services/task/index.ts (hunk 25 above)
Comment (Kesari3008) [reply packages/@webex/plugin-cc/src/services/task/index.ts]: yes that's alright

**Review by Kesari3008 — APPROVED** [review]

**Comment by adhmenon** [comment]
Merged with main feature branch and tested all cases once again.
Everything is working as usual - [Vidcast](https://app.vidcast.io/share/cba23197-a8f4-4339-9251-9667ccd58b53)
